### PR TITLE
ENH: `stats.differential_entropy`: add array API support

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
Toward gh-20544

#### What does this implement/fix?
Adds array API support to `differential_entropy`. 

#### Additional information
Turned out pretty clean, I think.

### CODE REVIEW

File: scipy/stats/_entropy.py [hunk 1]
```diff
@@ -353,74 +354,73 @@ class of statistics based on spacings. Scandinavian Journal of
...
     # No equation number, but referred to as HVE_mn.
     # Typo: there should be a log within the summation.
     n = X.shape[-1]
     difference = X[..., m:] - X[..., :-m]
-    term1 = 1/(n-m) * np.sum(np.log((n+1)/m * difference), axis=-1)
-    k = np.arange(m, n+1)
-    return term1 + np.sum(1/k) + np.log(m) - np.log(n+1)
+    term1 = 1/(n-m) * xp.sum(xp.log((n+1)/m * difference), axis=-1)
+    k = xp.arange(m, n+1.)
+    return term1 + xp.sum(1./k) + math.log(m) - math.log(n+1)
```
Comment (mdhaber) [inline scipy/stats/_entropy.py:391]:
```suggestion
    k = xp.arange(m, n+1, dtype=term1.dtype)
    return term1 + xp.sum(1/k) + math.log(m) - math.log(n+1)
```

File: scipy/stats/_entropy.py [hunk 2]
```diff
@@ -353,74 +354,73 @@ class of statistics based on spacings. Scandinavian Journal of
...
     # No equation number, but referred to as HE_mn
     n = X.shape[-1]
-    X = _pad_along_last_axis(X, m)
+    X = _pad_along_last_axis(X, m, xp=xp)
 
     differences = X[..., 2 * m:] - X[..., : -2 * m:]
 
-    i = np.arange(1, n+1).astype(float)
-    ci = np.ones_like(i)*2
+    i = xp.arange(1., n+1, dtype=X.dtype)
```
Comment (mdhaber) [inline scipy/stats/_entropy.py:402]:
```suggestion
    i = xp.arange(1, n+1, dtype=X.dtype)
```

File: scipy/stats/tests/test_entropy.py [hunk 3]
```diff
@@ -278,27 +271,41 @@ def test_norm_rmse_std(self, method, expected):
...
 
-    @pytest.mark.parametrize('method, expected',
-                             list(expon_rmse_std_cases.items()))
-    def test_expon_rmse_std(self, method, expected):
+    rmse_std_cases = {stats.norm: norm_rmse_std_cases,
+                      stats.expon: expon_rmse_std_cases}
+
+    @pytest.mark.parametrize('method', ['vasicek', 'van es', 'ebrahimi', 'correa'])
+    @pytest.mark.parametrize('dist', [stats.norm, stats.expon])
+    def test_norm_rmse_std(self, method, dist, xp):
```
Comment (mdhaber) [inline scipy/stats/tests/test_entropy.py:279]:
```suggestion
    def test_rmse_std(self, method, dist, xp):
```

**Comment by mdhaber** [comment]
Does anyone more familiar with typing know why mypy doesn't know about distributions?
```
scipy/stats/tests/test_entropy.py:274: error: Module has no attribute "norm"  [attr-defined]
scipy/stats/tests/test_entropy.py:275: error: Module has no attribute "expon"  [attr-defined]
```
We have a few `type: ignore[attr-defined]` for this reason, but I wonder if there's a way to fix it permanently.

**Review by j-bowhay — APPROVED** [review]
Thanks @mdhaber

**Pull request merged**